EVMC support

chfast opened the pull request:
- The meat is in https://github.com/ethereum/evmc/pull/41, here included as a git submodule in vendor (to be copy&paste later with some vendor magic, but submodule is good for now because the code under review and is going to change).
- Here we just use the Go API from evmc package. Some improvements can be made on both sides in the future.
- The minus of having clean Go API is that the state tests execution bumped from 19 s to 24 s.
- The geth+aleth-interpreter is running on the mainnet as "aleth" on https://ethstats.net.

## Usage:

- The path to EVMC VM DLL can be provided with `EVMC_PATH` env var.
- The EVMC options can be passed to the VM via `EVMC_OPTIONS` env var, e.g. `EVMC_OPTIONS='debug=on evm2wasm=true'`.

chfast commented:
cc @gballet (I don't have the rights to make you the reviewer).

chfast commented on core/vm/evmc.go:
```diff
@@ -0,0 +1,583 @@
+
+func loadVM(path string) {
+	if len(path) == 0 {
+		panic("EVMC path not provided, use --vm flag")
```
After hours of running, geth reported this error. Does geth restart itself?

chfast commented on core/vm/evmc.go:
```diff
@@ -0,0 +1,583 @@
+	defer C.free(unsafe.Pointer((centrypoint)))
+
+	C.dlerror()
+	createSymbol = C.dlsym(handle, centrypoint)
```
The `dl` library does not have any way to iterate over symbols. We'd have to use ELF library on Linux and something equivalent for MACHO on macOS. Possible in the future, but definitely overkill at the moment.

chfast commented on ethstats/ethstats.go:
```diff
@@ -380,11 +380,12 @@ func (s *Service) login(conn *websocket.Conn) error {
 		network = fmt.Sprintf("%d", infos.Protocols["les"].(*les.NodeInfo).Network)
 		protocol = fmt.Sprintf("les/%d", les.ClientProtocolVersions[0])
 	}
+	node := strings.Replace(infos.Name, "Geth", "Geth+EVMC", 1)
```
I could not find where the name comes from, but this is not going to stay.

axic commented:
```
core/vm/evmc.go:21:10: fatal error: 'evmc/evmc.h' file not found
#include <evmc/evmc.h>
         ^~~~~~~~~~~~~
```

Don’t seem to have a submodule, does this rely on having `evmc` installed system wide?

axic commented:
To compile: changed the include path in `core/vm/evmc.go` to `../../evmc/include` and added `evmc` as a submodule in the root directory.

axic commented:
One issue I find with this that the existence of the requested VM is not checked on startup, e.g. `geth --vm invalid` will not result in a failure until code is executed.

chfast commented:
> To compile: changed the include path in core/vm/evmc.go to ../../evmc/include and added evmc as a submodule in the root directory.

I will pull your change to make it easier for now. The final solution would be to include evmc Go wrapper in ethereum/evmc repo and use that as a Go package here. No git submodule would be needed then.

> One issue I find with this that the existence of the requested VM is not checked on startup, e.g. geth --vm invalid will not result in a failure until code is executed.

Yes. Command line arguments in geth are quite flaky and to pass the path to VM you would have to change ~20 places (not all of them are covered at the moment, I added some workarounds). That requires discussion with Go team.

chfast commented:
This is one type of crash I got after days of running:
```
runtime: pointer 0xc700000000 to unused region of span idx=0x170000 span.base()=0xc432afc000 span.limit=0xc432afdf80 span.state=1
fatal error: found bad pointer in Go heap (incorrect use of unsafe or cgo?)

runtime stack:
runtime.throw(0xf51699, 0x3e)
	/usr/lib/go-1.10/src/runtime/panic.go:616 +0x81 fp=0x7f4022ffcd88 sp=0x7f4022ffcd68 pc=0x440fe1
runtime.heapBitsForObject(0xc700000000, 0x0, 0x0, 0xc53ad77500, 0xc417294457, 0xc400000000, 0x7f3fc010d278, 0x54)
	/usr/lib/go-1.10/src/runtime/mbitmap.go:425 +0x473 fp=0x7f4022ffcde0 sp=0x7f4022ffcd88 pc=0x4284b3
runtime.wbBufFlush1(0xc420055900)
	/usr/lib/go-1.10/src/runtime/mwbbuf.go:234 +0xde fp=0x7f4022ffce60 sp=0x7f4022ffcde0 pc=0x43d0ee
runtime.wbBufFlush.func1()
	/usr/lib/go-1.10/src/runtime/mwbbuf.go:180 +0x3a fp=0x7f4022ffce78 sp=0x7f4022ffce60 pc=0x46bfaa
runtime.systemstack(0x0)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:409 +0x79 fp=0x7f4022ffce80 sp=0x7f4022ffce78 pc=0x46dd49
runtime.mstart()
	/usr/lib/go-1.10/src/runtime/proc.go:1175 fp=0x7f4022ffce88 sp=0x7f4022ffce80 pc=0x445620

goroutine 116 [running]:
runtime.systemstack_switch()
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:363 fp=0xc43a4095e0 sp=0xc43a4095d8 pc=0x46dcc0
runtime.wbBufFlush(0xc47394db58, 0x0)
	/usr/lib/go-1.10/src/runtime/mwbbuf.go:179 +0x4e fp=0xc43a409600 sp=0xc43a4095e0 pc=0x43cf7e
runtime.gcWriteBarrier(0x0, 0xc43a4096c8, 0x85e8be, 0x0, 0x0, 0x1b31318, 0x0, 0x0, 0xc43a4098e8, 0x85cdc2, ...)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:2442 +0xb4 fp=0xc43a409688 sp=0xc43a409600 pc=0x470914
runtime.deferreturn(0x7f4014001000)
	/usr/lib/go-1.10/src/runtime/panic.go:342 +0xc5 fp=0xc43a4096d8 sp=0xc43a409688 pc=0x440415
github.com/ethereum/go-ethereum/core/vm.(*EVMC).Run(0xc50241db40, 0xc43078dc70, 0xc4aaf96d70, 0x44, 0x44, 0x1b31318, 0x0, 0x0, 0x0, 0x0)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:624 +0x642 fp=0xc43a4098f8 sp=0xc43a4096d8 pc=0x85cdc2
github.com/ethereum/go-ethereum/core/vm.run(0xc4ec65e000, 0xc43078dc70, 0xc4aaf96d70, 0x44, 0x44, 0xd48, 0xd80, 0x57e6942623a0d061, 0xc438802e18, 0x2f928f9b87331201)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/vm/evm.go:52 +0x76 fp=0xc43a409968 sp=0xc43a4098f8 pc=0x842636
github.com/ethereum/go-ethereum/core/vm.(*EVM).Call(0xc4ec65e000, 0x10a4500, 0xc5f02e66e0, 0x9f12823e21e32505, 0xd030980e27d32c65, 0x4975e9c5, 0xc4aaf96d70, 0x44, 0x44, 0x348a4, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/vm/evm.go:189 +0x4f6 fp=0xc43a409b28 sp=0xc43a409968 pc=0x842f16
github.com/ethereum/go-ethereum/core.(*StateTransition).TransitionDb(0xc47d7393b0, 0x10b2a60, 0xc437ab91a0, 0xc5f387c130, 0xc47d7393b0, 0xc437ab91a0, 0xc437ab91a0, 0xee6b00)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/state_transition.go:213 +0x3c1 fp=0xc43a409d20 sp=0xc43a409b28 pc=0x883591
github.com/ethereum/go-ethereum/core.ApplyMessage(0xc4ec65e000, 0x10b2a60, 0xc437ab91a0, 0xc5f387c130, 0x3f1dee3a2fd16533, 0x310651c5, 0xc509061640, 0x0, 0x0, 0x0, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/state_transition.go:132 +0x5b fp=0xc43a409d70 sp=0xc43a409d20 pc=0x882c6b
github.com/ethereum/go-ethereum/core.ApplyTransaction(0x1794680, 0x10a6100, 0xc420d0e000, 0xc43a40f3c0, 0xc5f387c130, 0xc42087b420, 0xc562832d80, 0xc5129b2a20, 0xc562832f50, 0x0, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/core/state_processor.go:99 +0x29f fp=0xc43a40cab8 sp=0xc43a409d70 pc=0x88236f
github.com/ethereum/go-ethereum/miner.(*Work).commitTransaction(0xc54cbe4000, 0xc5129b2a20, 0xc420d0e000, 0x0, 0x0, 0x0, 0xc5f387c130, 0x0, 0x0, 0xc509cadd20, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:609 +0x13a fp=0xc43a40f3a8 sp=0xc43a40cab8 pc=0xa48dda
github.com/ethereum/go-ethereum/miner.(*Work).commitTransactions(0xc54cbe4000, 0xc4202cf9e0, 0xc5624604b0, 0xc420d0e000, 0x0, 0x0, 0x0)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:555 +0x297 fp=0xc43a40f748 sp=0xc43a40f3a8 pc=0xa48057
github.com/ethereum/go-ethereum/miner.(*worker).commitNewWork(0xc420123a20)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:455 +0xa2c fp=0xc43a40fd20 sp=0xc43a40f748 pc=0xa464cc
github.com/ethereum/go-ethereum/miner.(*worker).update(0xc420123a20)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:254 +0x781 fp=0xc43a40ffd8 sp=0xc43a40fd20 pc=0xa44b31
runtime.goexit()
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc43a40ffe0 sp=0xc43a40ffd8 pc=0x470851
created by github.com/ethereum/go-ethereum/miner.newWorker
	/home/chfast/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:158 +0x4ae
```

chfast commented:
This is a crash that can be easily reproduces with `GOCACHE=off go test ./tests -run TestState`:

```
runtime: pointer 0xc4212a4670 to unallocated span idx=0x952 span.base()=0xc4212a0000 span.limit=0xc4212b0000 span.state=3
fatal error: found bad pointer in Go heap (incorrect use of unsafe or cgo?)

runtime stack:
runtime.throw(0x8a1338, 0x3e)
	/usr/lib/go-1.10/src/runtime/panic.go:616 +0x81 fp=0x7f8b137fd248 sp=0x7f8b137fd228 pc=0x42dd31
runtime.heapBitsForObject(0xc4212a4670, 0x0, 0x0, 0xc421dd8c00, 0xc41ff1139f, 0xc400000000, 0x7f8b3c1f4250, 0x30)
	/usr/lib/go-1.10/src/runtime/mbitmap.go:425 +0x473 fp=0x7f8b137fd2a0 sp=0x7f8b137fd248 pc=0x4152b3
runtime.wbBufFlush1(0xc420041900)
	/usr/lib/go-1.10/src/runtime/mwbbuf.go:234 +0xde fp=0x7f8b137fd320 sp=0x7f8b137fd2a0 pc=0x429c5e
runtime.wbBufFlush.func1()
	/usr/lib/go-1.10/src/runtime/mwbbuf.go:180 +0x3a fp=0x7f8b137fd338 sp=0x7f8b137fd320 pc=0x4581aa
runtime.systemstack(0x7f8b137fd700)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:409 +0x79 fp=0x7f8b137fd340 sp=0x7f8b137fd338 pc=0x459c19
runtime.mstart()
	/usr/lib/go-1.10/src/runtime/proc.go:1175 fp=0x7f8b137fd348 sp=0x7f8b137fd340 pc=0x432370

goroutine 7747 [running, locked to thread]:
runtime.systemstack_switch()
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:363 fp=0xc4225abef8 sp=0xc4225abef0 pc=0x459b90
runtime.wbBufFlush(0xc422056818, 0x0)
	/usr/lib/go-1.10/src/runtime/mwbbuf.go:179 +0x4e fp=0xc4225abf18 sp=0xc4225abef8 pc=0x429aee
runtime.gcWriteBarrier(0xc4c7a6a6ae7b5e00, 0x26c3fa7e97ebdfc2, 0xc4872d55af, 0xc4c7a6a6ae7b5e09, 0x26c3fa7e97ebdfc2, 0x872d55af, 0xc42208b6c0, 0xc4c7a6a6ae7b5e09, 0x3, 0xc420041970, ...)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:2442 +0xb4 fp=0xc4225abfa0 sp=0xc4225abf18 pc=0x45c7c4
runtime.freedefer(0xc422056800)
	/usr/lib/go-1.10/src/runtime/panic.go:283 +0x1a0 fp=0xc4225ac030 sp=0xc4225abfa0 pc=0x42cd80
runtime.deferreturn(0xc4225ac0c7)
	/usr/lib/go-1.10/src/runtime/panic.go:344 +0x82 fp=0xc4225ac080 sp=0xc4225ac030 pc=0x42cf42
runtime.cgocallbackg1(0x0)
	/usr/lib/go-1.10/src/runtime/cgocall.go:332 +0x1a7 fp=0xc4225ac100 sp=0xc4225ac080 pc=0x404717
runtime.cgocallbackg(0x0)
	/usr/lib/go-1.10/src/runtime/cgocall.go:194 +0xda fp=0xc4225ac168 sp=0xc4225ac100 pc=0x4044da
runtime.cgocallback_gofunc(0x404388, 0x790ca0, 0xc4225ac210, 0x7f8b3ee426c8)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:826 +0x9b fp=0xc4225ac188 sp=0xc4225ac168 pc=0x45b52b
runtime.asmcgocall(0x790ca0, 0xc4225ac210)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:673 +0x42 fp=0xc4225ac190 sp=0xc4225ac188 pc=0x45b3c2
runtime.cgocall(0x790ca0, 0xc4225ac210, 0x4138e8)
	/usr/lib/go-1.10/src/runtime/cgocall.go:131 +0x88 fp=0xc4225ac1c8 sp=0xc4225ac190 pc=0x404388
github.com/ethereum/go-ethereum/core/vm._Cfunc_evmc_execute(0x7f8b3c1613a0, 0xc421238350, 0xc400000004, 0xc421464ab0, 0x7f8b080025a0, 0x2d, 0x0, 0x0, 0x0, 0x0, ...)
	_cgo_gotypes.go:283 +0x65 fp=0xc4225ac210 sp=0xc4225ac1c8 pc=0x754185
github.com/ethereum/go-ethereum/core/vm.(*EVMC).Run.func5(0x7f8b3c1613a0, 0xc421238350, 0x54dbd3e600000004, 0xc421464ab0, 0x7f8b080025a0, 0x2d, 0x0, 0x0, 0x0, 0x0, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:599 +0x1c1 fp=0xc4225ac2d0 sp=0xc4225ac210 pc=0x759671
github.com/ethereum/go-ethereum/core/vm.(*EVMC).Run(0xc4222ea580, 0xc420d4ab60, 0xd43bf0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:599 +0x542 fp=0xc4225ac4f0 sp=0xc4225ac2d0 pc=0x757bb2
github.com/ethereum/go-ethereum/core/vm.run(0xc421d1ea80, 0xc420d4ab60, 0xd43bf0, 0x0, 0x0, 0x2d, 0x2d, 0xb3f60b7a8cc9952f, 0x8, 0xc4212a4600)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evm.go:52 +0x76 fp=0xc4225ac560 sp=0xc4225ac4f0 pc=0x73cc76
github.com/ethereum/go-ethereum/core/vm.(*EVM).StaticCall(0xc421d1ea80, 0x8fcba0, 0xc420d4aa90, 0xc4c7a6a6ae7b5e19, 0x26c3fa7e97ebdfc2, 0x872d55af, 0xd43bf0, 0x0, 0x0, 0xd303e, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evm.go:310 +0x2f0 fp=0xc4225ac648 sp=0xc4225ac560 pc=0x73e9e0
github.com/ethereum/go-ethereum/core/vm.call(0x7f8b137fd840, 0xc421238300, 0x7f8b137fd880)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:472 +0x15f fp=0xc4225ac780 sp=0xc4225ac648 pc=0x756f1f
github.com/ethereum/go-ethereum/core/vm._cgoexpwrap_2150e86e3465_call(0x7f8b137fd840, 0xc421238300, 0x7f8b137fd880)
	_cgo_gotypes.go:473 +0x3f fp=0xc4225ac7a8 sp=0xc4225ac780 pc=0x754bcf
runtime.call32(0x0, 0x7f8b137fd750, 0x7f8b137fd7e0, 0x18)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:573 +0x3b fp=0xc4225ac7d8 sp=0xc4225ac7a8 pc=0x459f5b
runtime.cgocallbackg1(0x0)
	/usr/lib/go-1.10/src/runtime/cgocall.go:316 +0x19c fp=0xc4225ac858 sp=0xc4225ac7d8 pc=0x40470c
runtime.cgocallbackg(0x0)
	/usr/lib/go-1.10/src/runtime/cgocall.go:194 +0xda fp=0xc4225ac8c0 sp=0xc4225ac858 pc=0x4044da
runtime.cgocallback_gofunc(0x404388, 0x790ca0, 0xc4225ac968, 0x7f8b3ee426c8)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:826 +0x9b fp=0xc4225ac8e0 sp=0xc4225ac8c0 pc=0x45b52b
runtime.asmcgocall(0x790ca0, 0xc4225ac968)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:673 +0x42 fp=0xc4225ac8e8 sp=0xc4225ac8e0 pc=0x45b3c2
runtime.cgocall(0x790ca0, 0xc4225ac968, 0x4138e8)
	/usr/lib/go-1.10/src/runtime/cgocall.go:131 +0x88 fp=0xc4225ac920 sp=0xc4225ac8e8 pc=0x404388
github.com/ethereum/go-ethereum/core/vm._Cfunc_evmc_execute(0x7f8b3c1613a0, 0xc421238300, 0xc400000004, 0xc421464a20, 0x7f8b080024f0, 0x17, 0x0, 0x0, 0x0, 0x0, ...)
	_cgo_gotypes.go:283 +0x65 fp=0xc4225ac968 sp=0xc4225ac920 pc=0x754185
github.com/ethereum/go-ethereum/core/vm.(*EVMC).Run.func5(0x7f8b3c1613a0, 0xc421238300, 0x75679e3e00000004, 0xc421464a20, 0x7f8b080024f0, 0x17, 0x0, 0x0, 0x0, 0x0, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:599 +0x1c1 fp=0xc4225aca28 sp=0xc4225ac968 pc=0x759671
github.com/ethereum/go-ethereum/core/vm.(*EVMC).Run(0xc4222ea580, 0xc420d4aa90, 0xd43bf0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:599 +0x542 fp=0xc4225acc48 sp=0xc4225aca28 pc=0x757bb2
github.com/ethereum/go-ethereum/core/vm.run(0xc421d1ea80, 0xc420d4aa90, 0xd43bf0, 0x0, 0x0, 0x17, 0x17, 0x766a7f087c215cc1, 0x82dd01, 0xc4201fb8c0)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evm.go:52 +0x76 fp=0xc4225accb8 sp=0xc4225acc48 pc=0x73cc76
github.com/ethereum/go-ethereum/core/vm.(*EVM).CallCode(0xc421d1ea80, 0x8fcba0, 0xc420d4a8f0, 0xc4c7a6a6ae7b5e09, 0x26c3fa7e97ebdfc2, 0x872d55af, 0xd43bf0, 0x0, 0x0, 0xd6981, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evm.go:234 +0x369 fp=0xc4225acde8 sp=0xc4225accb8 pc=0x73df29
github.com/ethereum/go-ethereum/core/vm.call(0x7f8b137fdc00, 0xc4212382a0, 0x7f8b137fdc40)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:480 +0x6f6 fp=0xc4225acf20 sp=0xc4225acde8 pc=0x7574b6
github.com/ethereum/go-ethereum/core/vm._cgoexpwrap_2150e86e3465_call(0x7f8b137fdc00, 0xc4212382a0, 0x7f8b137fdc40)
	_cgo_gotypes.go:473 +0x3f fp=0xc4225acf48 sp=0xc4225acf20 pc=0x754bcf
runtime.call32(0x0, 0x7f8b137fdb10, 0x7f8b137fdba0, 0x18)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:573 +0x3b fp=0xc4225acf78 sp=0xc4225acf48 pc=0x459f5b
runtime.cgocallbackg1(0x0)
	/usr/lib/go-1.10/src/runtime/cgocall.go:316 +0x19c fp=0xc4225acff8 sp=0xc4225acf78 pc=0x40470c
runtime.cgocallbackg(0x0)
	/usr/lib/go-1.10/src/runtime/cgocall.go:194 +0xda fp=0xc4225ad060 sp=0xc4225acff8 pc=0x4044da
runtime.cgocallback_gofunc(0x404388, 0x790ca0, 0xc4225ad108, 0x7f8b3ee426c8)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:826 +0x9b fp=0xc4225ad080 sp=0xc4225ad060 pc=0x45b52b
runtime.asmcgocall(0x790ca0, 0xc4225ad108)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:673 +0x42 fp=0xc4225ad088 sp=0xc4225ad080 pc=0x45b3c2
runtime.cgocall(0x790ca0, 0xc4225ad108, 0x4138e8)
	/usr/lib/go-1.10/src/runtime/cgocall.go:131 +0x88 fp=0xc4225ad0c0 sp=0xc4225ad088 pc=0x404388
github.com/ethereum/go-ethereum/core/vm._Cfunc_evmc_execute(0x7f8b3c1613a0, 0xc4212382a0, 0xc400000004, 0xc421464990, 0x7f8b080025e0, 0x2d, 0x0, 0x0, 0x0, 0x0, ...)
	_cgo_gotypes.go:283 +0x65 fp=0xc4225ad108 sp=0xc4225ad0c0 pc=0x754185
github.com/ethereum/go-ethereum/core/vm.(*EVMC).Run.func5(0x7f8b3c1613a0, 0xc4212382a0, 0x54dbd3e600000004, 0xc421464990, 0x7f8b080025e0, 0x2d, 0x0, 0x0, 0x0, 0x0, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:599 +0x1c1 fp=0xc4225ad1c8 sp=0xc4225ad108 pc=0x759671
github.com/ethereum/go-ethereum/core/vm.(*EVMC).Run(0xc4222ea580, 0xc420d4a8f0, 0xd43bf0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evmc.go:599 +0x542 fp=0xc4225ad3e8 sp=0xc4225ad1c8 pc=0x757bb2
github.com/ethereum/go-ethereum/core/vm.run(0xc421d1ea80, 0xc420d4a8f0, 0xd43bf0, 0x0, 0x0, 0x2d, 0x2d, 0xb3f60b7a8cc9952f, 0xc4215554a8, 0xda79474af2d88801)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evm.go:52 +0x76 fp=0xc4225ad458 sp=0xc4225ad3e8 pc=0x73cc76
github.com/ethereum/go-ethereum/core/vm.(*EVM).Call(0xc421d1ea80, 0x8fd5a0, 0xc420991a00, 0xc4c7a6a6ae7b5e19, 0x26c3fa7e97ebdfc2, 0x872d55af, 0xd43bf0, 0x0, 0x0, 0xef038, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/vm/evm.go:189 +0x4f6 fp=0xc4225ad618 sp=0xc4225ad458 pc=0x73d556
github.com/ethereum/go-ethereum/core.(*StateTransition).TransitionDb(0xc42186a230, 0x902200, 0xc421775560, 0xc422481d20, 0xc42186a230, 0x10, 0xc42258b001, 0xc422481d30)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/state_transition.go:213 +0x3c1 fp=0xc4225ad810 sp=0xc4225ad618 pc=0x77b281
github.com/ethereum/go-ethereum/core.ApplyMessage(0xc421d1ea80, 0x902200, 0xc421775560, 0xc422481d20, 0x1, 0xc422481d30, 0x0, 0x1, 0xc28fac6d66bce6e0, 0xbaf97f69, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/core/state_transition.go:132 +0x5b fp=0xc4225ad860 sp=0xc4225ad810 pc=0x77a95b
github.com/ethereum/go-ethereum/tests.(*StateTest).Run(0xc42208aee0, 0xc422481831, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/tests/state_test_util.go:143 +0x483 fp=0xc4225b04e0 sp=0xc4225ad860 pc=0x785473
github.com/ethereum/go-ethereum/tests.TestState.func1.1.1(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/tests/state_test.go:51 +0xb5 fp=0xc4225b2d78 sp=0xc4225b04e0 pc=0x78d855
github.com/ethereum/go-ethereum/tests.withTrace(0xc4213b0a50, 0xf4240, 0xc4225b7f50)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/tests/state_test.go:63 +0x67 fp=0xc4225b7ee8 sp=0xc4225b2d78 pc=0x78c1b7
github.com/ethereum/go-ethereum/tests.TestState.func1.1(0xc4213b0a50)
	/home/chfast/go/src/github.com/ethereum/go-ethereum/tests/state_test.go:50 +0x1ba fp=0xc4225b7fa8 sp=0xc4225b7ee8 pc=0x78da9a
testing.tRunner(0xc4213b0a50, 0xc4222ea040)
	/usr/lib/go-1.10/src/testing/testing.go:777 +0xd0 fp=0xc4225b7fd0 sp=0xc4225b7fa8 pc=0x4e3c10
runtime.goexit()
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc4225b7fd8 sp=0xc4225b7fd0 pc=0x45c701
created by testing.(*T).Run
	/usr/lib/go-1.10/src/testing/testing.go:824 +0x2e0
```

chfast commented:
Big update! See the top comment.

axic commented:
Needs rebase after #17093 was merged.

gballet commented on core/vm/evm.go:
```diff
@@ -139,7 +139,7 @@ func NewEVM(ctx Context, statedb StateDB, chainConfig *params.ChainConfig, vmCon
 	}
 
-	evm.interpreters[0] = NewEVMInterpreter(evm, vmConfig)
+	evm.interpreters[0] = NewEVMC(evm, vmConfig)
```
Since the merge of #17093, you can support other interpreters; this disables the evm interpreter which is not acceptable. What should happen is that if `--evmc-vm` is passed as an argument, then an EVMC interpreter should be inserted before. Something like:

```golang
if cfg.EVMCPath != nil {
  old := evm.interpreters
  evm.interpreters = make([]Interpreter, len(evm.interpreters)+1)
  evm.interpreters[0] = NewEVMC(evm, vmConfig)
  copy(evm.interpreters[1:], old)
}
```

gballet commented on core/vm/evmc.go:
```diff
@@ -0,0 +1,282 @@
+	return evmcInstance
+}
+
+func NewEVMC(env *EVM, cfg Config) *EVMC {
```
`EVMC` should implement the `Interpreter` interface, see https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L48

axic commented on .gitmodules:
```diff
@@ -1,3 +1,6 @@
 [submodule "tests"]
 	path = tests/testdata
 	url = https://github.com/ethereum/tests
+[submodule "evmc"]
```
EVMC is only the C header plus the Go binding for it, doesn't include any VM implementation. I think it is needed, due it being a compile time dependency?

gballet replied:
My bad, the previous comment was very poorly worded: you should use go's own opinionated package system and add the dependency to `vendor/vendor.json` with the `govendor` utility.

chfast commented on cmd/utils/flags.go:
```diff
@@ -355,6 +355,11 @@ var (
 		Usage: "Record information useful for VM and contract debugging",
 	}
+	EVMCPathFlag = cli.StringFlag{
+		Name:  "vm",
```
We can make it nicer by allowing options like:
```
--vm auto
--vm interpreter
--vm ewasm
--vm /path/to/evmc.so
```

This flag does not work anyway, so I'm removing it for now.

axic commented on core/vm/evmc.go:
```diff
@@ -0,0 +1,280 @@
+			output, gasLeftU, err = env.Call(host.contract, destination, input, gasU, value)
+		}
+	case evmc.DelegateCall:
+		output, gasLeftU, err = env.DelegateCall(host.contract, destination, input, gasU)
```
Should any of these panic if `static` is set or is it handled up in the interface?

chfast replied:
`static` is magically handled inside, but this seems to be bad design. Probably should be changed later on to be handled internally in any VM.

chfast commented:
@gballet I believe I addressed all your comments.

gballet commented on core/vm/evmc.go:
```diff
@@ -0,0 +1,307 @@
+	return evmc.StorageModified
+}
+
+func (host *HostContext) GetBalance(addr common.Address) common.Hash {
```
Rename to getExternalBalance

chfast replied:
It matches the EVMC naming: https://github.com/ethereum/evmc/blob/master/include/evmc/evmc.h#L485 and I want to match the names exactly.

I don't believe the "external" is needed here because it is implicit for all Host methods (all of them takes the address as an argument).

axic commented on core/vm/evm.go:
```diff
@@ -139,7 +140,11 @@ func NewEVM(ctx Context, statedb StateDB, chainConfig *params.ChainConfig, vmCon
 	}
 
-	evm.interpreters[0] = NewEVMInterpreter(evm, vmConfig)
+	if len(os.Getenv("EVMC_PATH")) != 0 {
```
What exactly is geth doing with that list?

chfast replied:
I tried the command line switch but this is not doable currently (at least not in this PR) because the `VMConfig{}` is created in 100 places usually with default value just because the it is required to create an EVM.

gballet replied:
@axic it lists the interpreters, so as to be able to find the best format for a given executable. This is how you can mix EVM&EWASM contracts.

chfast commented on core/vm/evmc.go:
```diff
@@ -0,0 +1,280 @@
+}
+
+func (evm *EVMC) CanRun([]byte) bool {
+	return true
```
So it should be `false` and then never run? EVMC is not for ewasm only, I'm using it with C++ EVM interpreter. I believe if an user specifies the execution as `geth --vm myvm.so` he really wants to use `myvm.so`.

You should rethink this `CanRun()` feature.

gballet replied:
the feature has been thought through enough, thank you. If you have `geth --vm myvm.so` you are stuck with only one interpreter, whereas this supports a mixed chain with different formats.

chfast replied:
This will be handled with EVMC 6.

gballet replied:
then in EVMC-6, and provided the right command line switch, the return value here should be true. Until then, it should check if the format is wasm and, if not, return false.

chfast replied:
This would break geth + aleth-interpreter.

gballet replied:
Please be more specific. How would that break geth + aleth ?

chfast replied:
I need geth + aleth-interpreter for testing. Aleth is EVM and if we check for wasm bytecode here it will be skip and the execution fallback to Go Interpreter.

chfast replied:
Anyway, I'm not aiming to merge this PR now. We should definitely wait for EVMC 6 which should be soon.

gballet replied:
ok let's do that. In the mean time, here is the PR that implements all the command line switches: https://github.com/ethereum/go-ethereum/pull/17687

chfast commented:
> Please update to EVMC 5.2.0.

Done.

> @chfast should support for EVMC_REJECTED implemented in this PR or a subsequent PR?

Let's wait for capabilities from EVMC 6.

axic commented:
Please rebase over #17687 (well, master) so that we can support fallback to EVM.

axic commented on core/vm/evmc.go:
```diff
@@ -0,0 +1,311 @@
+		defer func() { evm.readOnly = false }()
+	}
+
+	fmt.Printf("depth: %d kind: %d static: %t\n", evm.env.depth, kind, readOnly)
```
Debug to be removed?

chfast commented on core/vm/evmc.go:
```diff
@@ -0,0 +1,625 @@
+	"strings"
+)
+
+type EVMC struct {
```
It does implement the `Interpreter` interface.

chfast commented:
Replaced with #17954
